Add ambient temperature and humidity status sensors to NUT

tdfountain opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
Add ambient temperature and humidity status sensors. PDUs and select UPSs monitor the status of ambient environmental and report status relative to pre-defined ranges.  This status information is communicated with NUT using ambient.humidity.status and ambient.temperature.status.  These sensors are useful for signaling a warning or critical event to an environmental condition.  This information is Diagnostic in nature since it is not the primary purpose of a PDU or UPS.

Implementation uses translation keys with both strings.json and icons.json.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

tdfountain committed:
Add ambient temperature and humidity status sensors

davet2001 reviewed (changes requested):
Looks good to me. Thanks @tdfountain!

Are you able to give some examples of the sensor status? That might help allocate it to an enumerated data type if there is something already existing.

I assume it displays a text string.

tdfountain commented:
Thanks for reviewing.  I believe it is a text field returned by NUT.

At least for Eaton devices, these are mapped to MIB entries that track threshold statuses.  The two relevant thresholds status are reported as integer values humidityThStatus and temperatureThStatus  .  The corresponding MIB entry is:

```
humidityThStatus  OBJECT-TYPE
        SYNTAX     INTEGER {
                good (0),
                lowWarning (1),
                lowCritical (2),
                highWarning (3),
                highCritical (4)
        }
        MAX-ACCESS read-only
        STATUS     current
        DESCRIPTION 
                "Status of the measured humidity relative to the configured
                      thresholds."
```

and

```
temperatureThStatus  OBJECT-TYPE
        SYNTAX     INTEGER {
                good (0),
                lowWarning (1),
                lowCritical (2),
                highWarning (3),
                highCritical (4)
        }
        MAX-ACCESS read-only
        STATUS     current
        DESCRIPTION 
                "Status of the measured temperature relative to the configured
                      thresholds."
```

The NUT drivers I reviewed all map these to "good", "warning-low", "critical-low", "warning-high", and "critical-high".  It is possible some other NUT drivers map to different text strings but those are the ones I could find in the NUT source code.

The PDUs I have return "good" in a normal operational state.

davet2001 commented on homeassistant/components/nut/sensor.py:
```diff
@@ -928,13 +928,23 @@
         device_class=SensorDeviceClass.TEMPERATURE,
         state_class=SensorStateClass.MEASUREMENT,
     ),
+    "ambient.temperature.status": SensorEntityDescription(
+        key="ambient.temperature.status",
+        translation_key="ambient_temperature_status",
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
```
I don't think there is an existing enum that matches. However I think you can use `SensorDeviceClass.ENUM` here and use `options=` to list them.

That way the automation editor should see the options and the user will be able to select from a drop-down them e.g. to trigger automations.

This is better than just having raw text where the user has no idea of possible future values.

tdfountain replied:
Reviewed NUT source code to confirm this is an exhaustive list of currently supported return values.  Updated NUT component sensor class to use SensorDeviceClass of ENUM, set options, and wrote lambda function.

tdfountain committed:
Set ENUM as device class for humidity and temperature threshold status

tdfountain commented:
Please let me know if any additional information or changes are required. I have a number of other NUT improvements that I would like to commit but cannot until these are resolved. Thanks!

davet2001 commented:
@tdfountain if it is now ready, please make the PR as ready for review.

Thanks!

davet2001 commented:
@tdfountain Please can you increase the test coverage to enable the checks to pass?

tdfountain committed:
Add pytest case for no device to improve coverage

tdfountain committed:
Add pytest case for options update to improve coverage

tdfountain committed:
Add pytest case for PDUs, rename UPS test cases accordingly, add PDU fixture for testing, and add tests for ambient temperature and humidity status sensors

tdfountain committed:
Add pytest case for PDUs, rename UPS test cases accordingly, add PDU fixture for testing, and add tests for ambient temperature and humidity status sensors

tdfountain commented:
Added two pytest cases not related to this PR to improve code coverage and bring NUT component code coverage to 100%.  Added pytest and fixture to test for temperature and humidity status sensors.

The pytest code coverage failure appears to be due to other components.  Please let me know if I now need to go learn another component to be able to submit this PR.

bdraco commented on homeassistant/components/nut/sensor.py:
```diff
@@ -920,23 +936,39 @@
+        translation_key="ambient_temperature_status",
+        device_class=SensorDeviceClass.ENUM,
+        options=AMBIENT_THRESHOLD_STATUS_OPTIONS,
+        ambient_threshold_value_fn=lambda x: x.lower() if isinstance(x, str) else None,
```
Could be methodcaller instead of a lambda

bdraco replied:
context: I can't see how this can be anything other than a `str` based on https://github.com/bdraco/aionut/blob/main/src/aionut/client.py

bdraco replied:
If the value is missing than the sensor than available should return `False` and the state is never examined

tdfountain replied:
I can update.  I am not familiar with methodcaller though I have now read up on it a bit.  Is there a good example I should pattern match from since the powerwall component is the only one that uses methodcaller and it is a bit different.  Ideally one where the state is set to never examined.

tdfountain replied:
See comment below since lower() is not needed.

bdraco commented:
Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

bdraco commented on homeassistant/components/nut/const.py:
```diff
@@ -37,6 +37,14 @@
     "TEST": "Battery Testing",
 }
 
+AMBIENT_THRESHOLD_STATUS_OPTIONS = [
```
This looks specific to the sensor platform so it should stay there

tdfountain replied:
Agreed, moved.

tdfountain committed:
Move ambient constant to sensor platform since used exclusively by those functions

tdfountain committed:
Add @tdfountain to codeowners

bdraco commented:
```diff
diff --git a/homeassistant/components/nut/sensor.py b/homeassistant/components/nut/sensor.py
index 2bdfcdab5b4..e212cdf1ad3 100644
--- a/homeassistant/components/nut/sensor.py
+++ b/homeassistant/components/nut/sensor.py
@@ -6,7 +6,7 @@ from collections.abc import Callable
 from dataclasses import asdict, dataclass
 import logging
 from typing import Final, cast
-
+from operator import methodcaller
 from homeassistant.components.sensor import (
     SensorDeviceClass,
     SensorEntity,
@@ -951,7 +951,7 @@ SENSOR_TYPES: Final[dict[str, NutSensorEntityDescription]] = {
         translation_key="ambient_humidity_status",
         device_class=SensorDeviceClass.ENUM,
         options=AMBIENT_THRESHOLD_STATUS_OPTIONS,
-        ambient_threshold_value_fn=lambda x: x.lower() if isinstance(x, str) else None,
+        ambient_threshold_value_fn=methodcaller("lower"),
         entity_category=EntityCategory.DIAGNOSTIC,
     ),
     "ambient.temperature": NutSensorEntityDescription(
@@ -967,7 +967,7 @@ SENSOR_TYPES: Final[dict[str, NutSensorEntityDescription]] = {
         translation_key="ambient_temperature_status",
         device_class=SensorDeviceClass.ENUM,
         options=AMBIENT_THRESHOLD_STATUS_OPTIONS,
-        ambient_threshold_value_fn=lambda x: x.lower() if isinstance(x, str) else None,
+        ambient_threshold_value_fn=methodcaller("lower"),
         entity_category=EntityCategory.DIAGNOSTIC,
     ),
     "watts": NutSensorEntityDescription(
```

But you also need a new entity type that calls the callable.

tdfountain commented:
My apologies but I need a bit more help on this construct.  I originally made those changes but struggled with the new entity type.  I revised the NutSensorEntityDescription class to define the following:

`ambient_threshold_value_fn: Callable[[NutAmbientThresholdSensor], StateType]
`

and I tried creating a new entity:

```
@dataclass
class NutAmbientThresholdSensor(SensorEntityDescription):
    """Describes Nut ambient threshold sensor state."""
    def __init__(self, string: str) -> None:
        """Initialize the sensor."""
        self._string = string

    @property
    def native_value(self) -> str | None:
        """Get the current value."""
        return _string
```

The above unfortunately generates an error: "TypeError: non-default argument 'ambient_threshold_value_fn' follows default argument".  I have tried a number of other variants without success.  I was trying to mimic the powerwall component's MeterResponse but I am clearly missing something.  Thank you for any additional direction.

emontnemery commented:
> My apologies but I need a bit more help on this construct. I originally made those changes but struggled with the new entity type.

A custom `__init__` in a `dataclass` is probably not what you want. Please message me, @emontnemery, on Discord and let's try to work this out 👍

tdfountain committed:
Partial conversion from lambda to methodcaller

tdfountain committed:
Remove methodcaller since return value is lowercase

tdfountain committed:
Merge branch 'dev' into add-ambient-sensors-status

tdfountain commented:
The state string returned by NUT is always lowercase (the options are provided).  I therefore believe we do not need either a lambda function or a methodcaller.

davet2001 commented:
Just had a look through again - from my side all ok but I am not familiar with methodcaller etc so can't really comment on that.

Thanks @tdfountain!

bdraco commented on tests/components/nut/test_diagnostics.py:
```diff
@@ -41,3 +43,35 @@ async def test_diagnostics(
+        "commands": [],
+    }
+
+    with patch(
+        "homeassistant.helpers.device_registry.DeviceRegistry.async_get_device",
+        return_value=None,
+    ):
```
We shouldn't patch Home Assistant internals in tests.

tdfountain replied:
Fixed, removed patch and overrode unique_id instead to force error condition instead to improve test coverage.

bdraco commented on tests/components/nut/util.py:
```diff
@@ -79,3 +80,29 @@ async def async_init_integration(
     return entry
+
+
+async def _test_sensor_and_attributes(
```
If you want to assert outside of the test itself, you'll need to register a rewrite in `conftest.py` so these assertions get treated as test assertion

```
pytest.register_assert_rewrite("tests.components.nut.util")
```

tdfountain replied:
Fixed, added conftest.py.

bdraco committed:
Merge branch 'dev' into add-ambient-sensors-status

bdraco commented on homeassistant/components/nut/sensor.py:
```diff
@@ -46,22 +46,36 @@
-SENSOR_TYPES: Final[dict[str, SensorEntityDescription]] = {
-    "ups.status.display": SensorEntityDescription(
+
+@dataclass(frozen=True)
+class NutSensorEntityDescription(SensorEntityDescription):
+    """Describes Nut sensor entity."""
```
Since we don't have anything custom here now, can we drop this back to `SensorEntityDescription`?

tdfountain replied:
Thanks @davet2001 and @bdraco.  Fixed, class removed and reverted to standard SensorEntityDescirption.

bdraco commented on tests/components/nut/test_sensor.py:
```diff
@@ -92,6 +98,65 @@ async def test_devices_with_unique_ids(
+        hass,
+        entity_registry,
+        model,
+        unique_id=unique_id_base + "input.voltage",
```
Please use f-strings

tdfountain replied:
Fixed, change to f-string.

bdraco commented on tests/components/nut/test_sensor.py:
```diff
@@ -92,6 +98,65 @@ async def test_devices_with_unique_ids(
+        hass,
+        entity_registry,
+        model,
+        unique_id=unique_id_base + "ambient.humidity.status",
```
Please use f-strings

tdfountain replied:
Fixed, change to f-string.

bdraco commented on tests/components/nut/test_sensor.py:
```diff
@@ -92,6 +98,65 @@ async def test_devices_with_unique_ids(
+        hass,
+        entity_registry,
+        model,
+        unique_id=unique_id_base + "ambient.temperature.status",
```
Please use f-strings

tdfountain replied:
Fixed, change to f-string.

tdfountain committed:
Use f-strings for construction of unique ID

tdfountain committed:
Remove patch of Home Assistant helper and clear unique ID searched

tdfountain committed:
Register assert rewrite to allow factored testing

tdfountain committed:
Change from NutSensorEntityDescription to SensorEntityDescription since not customized anylonger

tdfountain committed:
Merge branch 'add-ambient-sensors-status' of https://github.com/tdfountain/home-assistant-core into add-ambient-sensors-status

tdfountain commented on tests/components/nut/test_diagnostics.py:
```diff
@@ -41,3 +41,31 @@ async def test_diagnostics(
+        list_commands_return_value=None,
+    )
+
+    mock_config_entry.runtime_data.unique_id = ""
```
This test case is to provide code coverage for an error condition in diagnostics.py at [Line 43](https://github.com/home-assistant/core/blob/dev/homeassistant/components/nut/diagnostics.py#L43).  The line should only occur when the result of device_registry.async_get_device() returns None.  I'm not seeing how to go about creating that condition "naturally" through the integration's setup.

The NUT component's setup uses the entry_id as the unique_id if there is no serial number.  The line in the PR was to basically keep the MockConfigEntry as valid and registered, but make it so that the device registry lookup fails.

This is unfortunately not code I wrote originally so I may well be missing something about when else this condition might occur.  I'm adding this test case just to address the automated code coverage comparison requirement.

Other suggestions for how to get device_registry.async_get_device() to return None?

bdraco replied:
AFAICT (mind you I've been on a plane for 9 hours today and little sleep so bear with me), that code on [line 43](https://github.com/home-assistant/core/blob/dev/homeassistant/components/nut/diagnostics.py#L43) is unreachable under normal circumstances because we unconditionally register the device at https://github.com/home-assistant/core/blob/db5bf417904a77fa2be75e555fac639400599b70/homeassistant/components/nut/__init__.py#L127

I suggest removing it and replacing it with an assertion.
```diff
diff --git a/homeassistant/components/nut/diagnostics.py b/homeassistant/components/nut/diagnostics.py
index 532e4ece76b..ec59fa65c22 100644
--- a/homeassistant/components/nut/diagnostics.py
+++ b/homeassistant/components/nut/diagnostics.py
@@ -39,8 +39,8 @@ async def async_get_config_entry_diagnostics(
     hass_device = device_registry.async_get_device(
         identifiers={(DOMAIN, hass_data.unique_id)}
     )
-    if not hass_device:
-        return data
+    # Device is always created
+    assert hass_device is not None
 
     data["device"] = {
         **attr.asdict(hass_device),
```

tdfountain replied:
Thanks for reviewing, that change makes sense to me.  Committed update to both diagnostics.py and associated pytest.

tdfountain committed:
Change to assert since dive is already created and remove associated pytest case

bdraco commented:
Thanks @tdfountain

tdfountain committed:
Merge branch 'dev' into add-ambient-sensors-status

bdraco committed:
Merge branch 'dev' into add-ambient-sensors-status
